src/driver.rs

```diff
@@ -410,6 +410,90 @@ impl DriverManager {
         Ok(Driver { c_driver })
     }
 
+    /// Get the Driver based on the file extension from filename
+    ///
+    /// Searches for the available extensions in the registered
+    /// drivers and returns the matches. The determined driver is
+    /// checked for writing capabilities as
+    /// [`Dataset::open`](Dataset::open) can already open datasets
+    /// with just path.
+    ///
+    /// See also: [`get_driver_by_name`](Self::get_driver_by_name)
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::DriverManager;
+    /// # fn main() -> gdal::errors::Result<()> {
+    /// let compatible_drivers =
+    ///     DriverManager::get_drivers_for_filename("test.gpkg", true)
+    ///         .iter()
+    ///         .map(|d| d.short_name())
+    ///         .collect::<Vec<String>>();
+    /// println!("{:?}", compatible_drivers);
+    /// # Ok(())
+    /// # }
+    /// ```
+    /// ```text
+    /// ["GPKG"]
+    /// ```
+    pub fn get_drivers_for_filename(filename: &str, is_vector: bool) -> Vec<Driver> {
+        let ext = {
+            let filename = filename.to_ascii_lowercase();
+            let e = match filename.rsplit_once('.') {
+                Some(("", _)) => "", // hidden file no ext
+                Some((f, "zip")) => {
+                    // zip files could be zipped shp or gpkg
+                    if f.ends_with(".shp") {
+                        "shp.zip"
+                    } else if f.ends_with(".gpkg") {
+                        "gpkg.zip"
+                    } else {
+                        "zip"
+                    }
+                }
+                Some((_, e)) => e, // normal file with ext
+                None => "",
+            };
+            e.to_string()
+        };
+
+        let mut drivers: Vec<Driver> = Vec::new();
+        for i in 0..DriverManager::count() {
+            let d = DriverManager::get_driver(i).expect("Index for this loop should be valid");
+            let can_create = d.metadata_item("DCAP_CREATE", "").is_some()
+                || d.metadata_item("DCAP_CREATECOPY", "").is_some();
+            let check_vector = is_vector && d.metadata_item("DCAP_VECTOR", "").is_some();
+            let check_raster = !is_vector && d.metadata_item("DCAP_RASTER", "").is_some();
+            let check_vector_translate =
+                is_vector && d.metadata_item("DCAP_VECTOR_TRANSLATE_FROM", "").is_some();
```

Comment:
See this description of what GDAL does when `Dataset::open` is called. 
https://github.com/OSGeo/gdal/blob/master/gcore/gdaldataset.cpp#L3630-L3645

From my 5 minute reading it looks like some of the logic is delegated to `GDALIdentifyDriverEx`:

https://github.com/OSGeo/gdal/blob/master/gcore/gdaldriver.cpp#L2649C25-L2649C45

My concern here is not wanting to diverge from the accepted GDAL logic.

Reply:
`GDALIdentifyDriverEx` is used for opening, but that's usually not necessary.

The functions I linked to in https://github.com/georust/gdal/issues/506#issuecomment-1868587093 is used by the GDAL binaries to pick an output format. If we offer this function, we should match the latter, and use a better name (`get_driver(s)_for_writing` or something like that).

Reply:
> The functions I linked to in https://github.com/georust/gdal/issues/506#issuecomment-1868587093 is used by the GDAL binaries to pick an output format.

Ahhh... I _thought_ there was something already in the code, but didn't realize it was C++ only :/

Reply:
Yeah, it is for binary apps. Not the part of the `gdal` library. Otherwise, I was planning on just using the function from `gdal-sys`.